### PR TITLE
soc : mec1501 : Deep and light sleep example

### PR DESCRIPTION
**Pull request by scottwcpg**
Created MEC1501 deep and light sleep example for MCHP MEC1501.
Modifications were made to SoC, board, timer, and hello world
sample program. Power management split into SoC power
implementing the interface and device power for device specific
logic.

Signed-off-by: Scott Worley <scott.worley@microchip.com>

### CODE REVIEW

File: soc/arm/microchip_mec/mec1501/device_power.c [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+ * Allow peripherals connected to external masters to wake the PLL but not
+ * the EC. Once the peripheral has serviced the external master the PLL
+ * will be turned back off. For example, if the eSPI master requests eSPI
+ * configuration information or state of virtual wires the EC doesn't need
+ * to be involved. The hardware can power on the PLL long enough to service
+ * the request and then turn the PLL back off.  The SMBus and I2C peripherals
+ * in slave mode can also make use of this feature.
+ */
+void soc_deep_sleep_non_wake_en(void)
+{
```
Comment (albertofloyd) [inline soc/arm/microchip_mec/mec1501/device_power.c]: I see PLL never shutting off when triggering deep sleep  from BIOS command line.
Better to have this disabled by default until root cause is known

**Comment by andrewboie** [comment]
@scottwcpg please rebase, CI failure seems unrelated

**Comment by scottwcpg** [comment]
Rebased per request.

File: soc/arm/microchip_mec/mec1501/power.c [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+	/* Mask all exceptions and interrupts except NMI and HardFault */
+	__set_PRIMASK(1);
+
+	soc_deep_sleep_periph_save();
+
+	soc_deep_sleep_enable();
+
+	soc_deep_sleep_wait_clk_idle();
+	soc_deep_sleep_non_wake_en();
```
Comment (albertofloyd) [inline soc/arm/microchip_mec/mec1501/power.c]: @scottwcpg  add macro to have this functionality disable by default

**Comment by albertofloyd** [comment]
@wentongwu  could you review this?

File: soc/arm/microchip_mec/mec1501/power.c [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+		break;
+	}
+}
+
+void _sys_pm_power_state_exit_post_ops(enum power_states state)
+{
+	switch (state) {
+#if (defined(CONFIG_SYS_POWER_SLEEP_STATES))
+	case SYS_POWER_STATE_SLEEP_1:
+		__enable_irq();
```
Comment (wentongwu) [inline soc/arm/microchip_mec/mec1501/power.c]: please use zephyr's interface to change irq state.

**Review by wentongwu — CHANGES_REQUESTED** [review]

File: soc/arm/microchip_mec/mec1501/power.c [hunk 4]
```diff
@@ -0,0 +1,134 @@
...
+	soc_deep_sleep_wait_clk_idle();
+	soc_deep_sleep_non_wake_en();
+
+	/*
+	 * Unmask all interrupts in BASEPRI. PRIMASK is used above to
+	 * prevent entering an ISR after unmasking in BASEPRI.
+	 * We clear PRIMASK in exit post ops.
+	 */
+	base_pri = __get_BASEPRI();
+	__set_BASEPRI(0);
```
[the remark this answers is not in the document]
Comment (scottwcpg) [reply soc/arm/microchip_mec/mec1501/power.c]: The Cortex-M4 will not wake on a pending interrupt if that interrupts priority is less than the mask value. If I use z_arch_irq_lock then the system will not wake due to the normal wake sources.  The point of setting PRIMASK=1 and BASEPRI=0 is to allow the peripheral interrupts to wake the system but NOT enter the ISR until all wake clean-up code which restores peripheral state. I looked but only found z_irq_lock/unlock which do not do what I need.

**Comment by albertofloyd** [comment]
@wentongwu could you review scott comments?

**Review by wentongwu — APPROVED** [review]

**Review by albertofloyd — APPROVED** [review]

**Review by dcpleung — APPROVED** [review]

**Review by tyrantlizard77 — APPROVED** [review]

File: soc/arm/microchip_mec/mec1501/power.c [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+#include <soc.h>
+#include "device_power.h"
+
+#if defined(CONFIG_SYS_POWER_DEEP_SLEEP_STATES)
+
+/*
+ * Deep Sleep
+ * Pros:
+ * Lower power dissipation, 48MHz PLL is off
+ * Cons:
```
Comment (wentongwu) [inline soc/arm/microchip_mec/mec1501/power.c]: so for deep sleep and light sleep, the difference is that the PLL(48Mhz) for CPU clock is on or off? If yes, can I understand that CPU just have different clock for deep/light sleep? @scottwcpg

File: soc/arm/microchip_mec/mec1501/device_power.c [hunk 6]
```diff
@@ -0,0 +1,219 @@
...
+			 | PCR_REGS->CLK_REQ2 | PCR_REGS->CLK_REQ3
+			 | PCR_REGS->CLK_REQ4;
+	} while ((clkreq != (1ul << MCHP_PCR1_CPU_POS)) && (cnt-- != 0));
+#endif
+}
+
+
+/*
+ * Allow peripherals connected to external masters to wake the PLL but not
+ * the EC. Once the peripheral has serviced the external master the PLL
```
Comment (wentongwu) [inline soc/arm/microchip_mec/mec1501/device_power.c]: from this comment, it seems there is another PLL for peripherals?

**Pull request merged**